### PR TITLE
Initial storage implementation.

### PR DESCRIPTION
**Pull request by tslamic**

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import (
+	"cloud.google.com/go/firestore"
+	"context"
+	"github.com/tslamic/go-oauth2-firestore/token"
+	"time"
+)
+
+type fstore struct {
+	c *firestore.Client
+	n string
```
Comment (stabenfeldt) [inline client.go:12]: What does `n` represent? Perhaps we could give it a more descriptive name, so one does not have to scroll further to see what it's used for?

File: client.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	ctx, cancel := context.WithTimeout(context.Background(), f.t)
+	defer cancel()
+	_, _, err := f.c.Collection(f.n).Add(ctx, token)
+	return err
+}
+
+func (f *fstore) Get(key string, val interface{}) (*token.Info, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), f.t)
+	defer cancel()
+	iter := f.c.Collection(f.n).Where(key, "==", val).Limit(1).Documents(ctx)
```
Comment (stabenfeldt) [inline client.go:26]: The key we're fetching the document by, is that also the ID? 
If so, perhaps we could just fetch it directly? 
e.g
`dsnap, err := client.Collection("cities").Doc("SF").Get(ctx)`

Then we could also skip the inter.Next() part.

https://firebase.google.com/docs/firestore/query-data/get-data

File: client.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	}
+	info := &token.Info{}
+	err = doc.DataTo(info)
+	return info, err
+}
+
+func (f *fstore) Del(key string, val interface{}) error {
+	ctx, cancel := context.WithTimeout(context.Background(), f.t)
+	defer cancel()
+	iter := f.c.Collection(f.n).Where(key, "==", val).Limit(1).Documents(ctx)
```
Comment (stabenfeldt) [inline client.go:39]: Same here. Can we fetch it by ID?

File: token/tokens.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
+package token
```
Comment (stabenfeldt) [inline token/tokens.go:1]: Perhaps we could split this into **AccessToken**, **RefreshToken**, etc?
To keep things short and sweet. :-)

**Review by stabenfeldt — CHANGES_REQUESTED** [review]

File: client.go (hunk 2 above)
Comment (tslamic) [reply client.go:26]: Spot the `Where` - we don't have the document ID, only the key and value we need to search for.

File: client.go (hunk 1 above)
Comment (tslamic) [reply client.go:12]: It's a collection name, wanted to keep it short and sweet, same as `c`, and though `collection` would be too long. Is `name` descriptive enough?

File: token/tokens.go (hunk 4 above)
Comment (tslamic) [reply token/tokens.go:1]: This is a library interface impl., I'd leave it as-is.

File: client.go (hunk 3 above)
Comment (tslamic) [reply client.go:39]: Same as above.

File: client.go (hunk 1 above)
Comment (stabenfeldt) [reply client.go:12]: `name` is better, but I would prefer `collectionName` even more.
That way we're able to keep the mental overhead of reading the code as low as possible.
When you read the name you don't have to think for a split second what it represents.

File: client.go (hunk 2 above)
Comment (stabenfeldt) [reply client.go:26]: 👍

**Review by stabenfeldt — CHANGES_REQUESTED** [review]
The build fails, BTW.

File: client.go (hunk 1 above)
Comment (tslamic) [reply client.go:12]: I think I'll keep it short & sweet, the mental overhead is negligible with 50 lines of code.

**Pull request merged**